cunumeric/array.py

```diff
@@ -955,12 +955,101 @@ def byteswap(self, inplace=False):
             numpy_array = self.__array__(stacklevel=3).byteswap(inplace=False)
             return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
 
-    @unimplemented
-    def choose(self, choices, out, mode="raise"):
-        numpy_array = self.__array__(stacklevel=3).choose(
-            choices=choices, out=out, mode=mode
-        )
-        return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
+    def choose(self, choices, out=None, mode="raise", stacklevel=2):
+        a = self.convert_to_cunumeric_ndarray(self)
+        if out is not None:
+            out = out.convert_to_cunumeric_ndarray(out)
+        # choices can be passed as array or tuple of arrays.
+        # in case of the tuple, we will pass each array separately
+        is_tuple = isinstance(choices, tuple)
+        if is_tuple:
```

Comment:
I think the logic would be cleaner from this point forward if you handled the case of `not isinstance(choices, tuple)` at this point, by extracting each row to a separate array, and creating a tuple:
```
if not isinstance(choices, tuple):
  choices = tuple(self.convert_to_cunumeric_ndarray(choices[i,...]) for i in range(n))
```
Then you could potentially drop the `args` and `is_tuple` path. Do you expect this to have any impact on performance? I haven't looked at the C++ level implementation yet.

Reply:
I can try this, but, I believe, it might be better to do opposite:
1)In case of the tuple: convert all subarrays to ndarray and concatenate them together into a single `choices` ndarray
2) in case of not tuple: just convert choices to ndarray. 
I think this should have better performance since we won't need to do [this ](https://github.com/ipdemes/cunumeric/blob/choose_operation/src/cunumeric/index/choose.cu#L55-L60)for passing vector of accessors to GPU

I couldn't implement this because `concatenate` method is not available in cunumeric quite yet.

Reply:
According to the discussion above, I think your approach might be better since it will avoid broadcasting choices.

Reply:
Also, concatenating arrays requires copying, whereas taking row-slices just creates a view. And passing the choices as a numpy array seems to be discouraged anyway. So let's go with always falling back to a tuple of choices.